Codetab

brandesNW opened the pull request:
Codetab ready for review.

Documentation will need some updates. 
There are plenty of comments indicating the underlying logic and use of the new classes, and many of the methods.

Some remarks with AAB indicate either my questions, or some small amount of work to be done.

example where trivial work is needed:
 netlogo-gui/src/main/sdm/gui/AggregateModelEditor.scala

Note: I haven't tested the api-ish methods in the last section of AbstractCodeTabPanel

LaCuneta reviewed (changes requested):
Congrats on getting things this far!   I think this is getting close.  I've added particular comments to the code.  Here are some general things:

- I've added notes where I see unsigned code comments, apologies if those are old comments you didn't make and not new ones, it's hard to tell with so much code moving around.  Feel free to ignore if you didn't add them.
- There is a commit to update the View2.5D extension.  I don't think that should be a part of this pull request.
- Because LevelSpace is bundled, the automated tests are not running (they fail at the build step).  LevelSpace will need to be updated for those tests to run and for this to be merged.  You can update LS on a branch and add a commit here to use that branch just to get things going, even if LS isn't final yet.
- If you have big blocks of explanatory comments for a class, I'd prefer those at the top of the class file rather than broken out into bits throughout the file.  I think `AppTabsManager.scala` could use this, but maybe a couple of the others as well.

I would overall prefer things to be more encapsulated - a single tabs manager class that handles everything seamlessly to the outside components - the rest of NetLogo GUI itself and also extensions that might want to add tabs.  Basically, nothing outside the tabs manager should ever have to worry about a tab index, a tab count, or about a particular tab that they didn't specifically add.  However, I understand that might be a bit too much of change for this go around, or might not be desirable for reasons I'm not thinking of.

brandesNW commented on netlogo-gui/src/main/app/AbstractTabsPanel.scala:
```diff
@@ -0,0 +1,175 @@
+import org.nlogo.app.interfacetab.InterfaceTab
+import org.nlogo.window.{ GUIWorkspace }
+
+// AbstractTabsPanel contains functionality common to Tabs and CodeTabsPanel
```
Please clarify the policy on initializing and dating comments
Are you saying when I create a new class,  I need to add my initials to every comment line that I add?
I thought the initials were for commenting places were you made changes to existing code that needed explanation.

If I don't change part of a file, but comment it to make what it does clearer, then should I add my initials?

I am happy to make the changes, but want to make sure I understand, so I get them all when I go through the file.

brandesNW replied:
Is there a workflow you use for making use of GitHub in doing updates - and then pulling it back to the local repo, or do you just make changes locally?

brandesNW replied:
I'm just going to sign all my comments etc. locally.

LaCuneta replied:
Because code comments have a tendency to get stale, I think it's a good idea to sign and date all of them.  It's just easier when reading code to know how recently a comment was made.  If a comment is too small to have a signature and date, it might best be removed.

You can make updates locally and push new commits to the branch when you're ready.  That will automatically update the PR.

brandesNW commented on netlogo-gui/src/main/app/AbstractTabsPanel.scala:
```diff
@@ -0,0 +1,175 @@
+
+// AbstractTabsPanel contains functionality common to Tabs and CodeTabsPanel
+
+abstract class AbstractTabsPanel(val workspace:           GUIWorkspace,
```
got it

brandesNW commented on netlogo-gui/src/main/app/AppTabManager.scala:
```diff
@@ -0,0 +1,340 @@
+    }
+  }
+
+  def getAppTabsOwner = { appTabsPanel }
```
I have some get functions such as getAppTabsOwner that are used by other classes to access members of a class.
In this case AbstractTabsPanel uses appTabsPanel. 
As you noted elsewhere, it might be possible to simplify some of this, but the fact that many objects in NetLogo have their state changed in multiple steps after they are first constructed makes things quite ticklish.

LaCuneta replied:
I'm still not clear on how it's useful to expose `appTabsPanel` as a method in addition to it being a `val` on the class (which will never change), but now `getAppTabsPanel` has also been added which again seems to do the exact same thing?  At a minimum there should only be one of these methods, and I think just letting people access `appTabsPanel` directly is fine.

brandesNW commented on netlogo-gui/src/main/app/AppTabManager.scala:
```diff
@@ -0,0 +1,340 @@
+    appTabsPanel.setSelectedIndex(index)
+  }
+
+  def getSelectedAppTabComponent() = { appTabsPanel.getSelectedComponent }
```
no, I can remove it

brandesNW commented on netlogo-gui/src/main/app/AppTabManager.scala:
```diff
@@ -0,0 +1,340 @@
+// to deal with the fact that CodeTabsPanel is instatiated only some of the time.
+
+class AppTabManager( val appTabsPanel:          Tabs,
+                     var codeTabsPanelOption: Option[CodeTabsPanel]) {
```
I called it that originally, then I called the thing you got when the option was Some realCodeTabsPanel.  For me this made the use of methods/code that are/is used only when CodeTabsPanel exists easier to understand.

I get the general principle, but sometimes I think making distinctions is useful, for example if you are dealing with different representations of keys.

Sometimes a unique name can also make searching the code easier.

brandesNW commented on netlogo-gui/src/main/app/codetab/MainCodeTab.scala:
```diff
@@ -41,7 +52,9 @@ with WindowEvents.LoadModelEvent.Handler
     // hack, to get it to realize it's really checked. ~Forrest (10/23/2007)
     smartTabAction.actionPerformed(null)
-    Seq(tabbing)
+    // The state of the check box is set in Tabs.handle(e: AfterLoadEvent)
```
From here on a blank comment field just means I will do as you suggest.

brandesNW commented on netlogo-gui/src/main/window/LinkRoot.scala:
```diff
@@ -5,5 +5,6 @@ package org.nlogo.window
 trait LinkRoot {
   protected val linkComponents = new collection.mutable.ListBuffer[AnyRef]()
   def addLinkComponent(c: AnyRef) { linkComponents += (c) }
+  def removeLinkComponent(c: AnyRef) { linkComponents += (c) }
```
good catch, a copy and paste error.

LaCuneta commented on netlogo-gui/src/main/app/AbstractTabsPanel.scala:
```diff
@@ -0,0 +1,175 @@
+    *
+    * @param tab The Component to remove.
+    */
+   def removeTab(tab: Component): Unit = {
```
One thing I forget to add - should this and `addNewTab()` both use `ExtendedCodeTab` as well, if they are meant to be the public interface?  Or am I missing something?

brandesNW replied:
The issue with replaceTab is that only a tab  of a similar type can be replaced.

You are right that to make everything more bullet-proof they  should work on an ExtendedTab class, and most methods could be made private[app]. 

I think introducing ExtendedTab is a good idea. 
I think it might be better to hold off on making things private.

brandesNW commented on netlogo-gui/src/main/app/App.scala:
```diff
@@ -41,6 +41,19 @@ import scala.io.Codec
  * for example code.
  */
 object App{
+  def printSwingObject(obj: Object, description: String): Unit = {
```
I will prefix all debug methods with __, and move all the ones in AppTabManager to the bottom of the class. 
The method printSwingObject in in the object, and is more general than the others. Are you ok w. moving it to the bottom of the object?

brandesNW commented on netlogo-gui/src/main/app/AppTabManager.scala:
```diff
@@ -0,0 +1,340 @@
+    if (tab.isInstanceOf[CodeTab]) getCodeTabsOwner else appTabsPanel
+  }
+
+  def isCodeTabSeparate = {
```
yes - inside the class it could but outside tabManager.isCodeTabSeparate is better than tabManager.codeTabsPanelOption.isDefined

so how about `def isCodeTabSeparate = { codeTabsPanelOption.isDefined }`

LaCuneta replied:
:+1: Yep, that's what I was meaning to suggest.

brandesNW commented on netlogo-gui/src/main/app/AppTabManager.scala:
```diff
@@ -0,0 +1,340 @@
+  // tabIndex = the index of the tab in tabOwner.
+  // This method allows for the possibility that the appTabsPanel has no tabs,
+  // although this should not occur in practice
+  def ownerAndIndexFromCombinedIndex(combinedTabIndx: Int): (AbstractTabsPanel, Int) = {
```
ok

brandesNW commented on netlogo-gui/src/main/app/AppTabManager.scala:
```diff
@@ -0,0 +1,340 @@
+    if (combinedTabIndx < 0) {
+      throw new IndexOutOfBoundsException
+    }
+    var tabOwner = appTabsPanel.asInstanceOf[AbstractTabsPanel]
```
done

brandesNW commented on netlogo-gui/src/main/app/AppTabManager.scala:
```diff
@@ -0,0 +1,340 @@
+    // check if it can refer to the a separate code tab
+    if (combinedTabIndx >= appTabCount) {
+      codeTabsPanelOption match {
+        case None           => throw new IndexOutOfBoundsException
```
The problem is just like that for any JTabbedPane that gets an invalid index. I'm not sure what text would clarify it. A suggestion is welcome.
If internal code is correct, it should never happen.
I don't see any way to make it NetLogo user friendly.
The stack trace will lead a developer to quickly grasp the problem.

brandesNW commented on netlogo-gui/src/main/app/AppTabManager.scala:
```diff
@@ -0,0 +1,340 @@
+        }
+      }
+    }
+    (tabOwner, tabIndex)
```
done.
some code standards are anti- multiple returns
I'm for them if they make the code clearer

brandesNW commented on netlogo-gui/src/main/app/AppTabManager.scala:
```diff
@@ -0,0 +1,340 @@
+  // Returns (null, -1) if there is no tab owner for this tab component.
+  def ownerAndIndexOfTab(tab: Component): (AbstractTabsPanel, Int) = {
+    var tabOwner = null.asInstanceOf[AbstractTabsPanel]
+    var tabIndex = appTabsPanel.indexOfComponent(tab)
```
done

brandesNW commented on netlogo-gui/src/main/app/AppTabManager.scala:
```diff
@@ -0,0 +1,340 @@
+  }
+
+  // Useful for debugging
+  def printAllTabs(): Unit = {
```
at the end of the file with a __

brandesNW commented on netlogo-gui/src/main/app/AppTabManager.scala:
```diff
@@ -0,0 +1,340 @@
+  }
+
+  // Useful for debugging
+  def printTabsOfTabsPanel(pane: JTabbedPane): Unit = {
```
done

brandesNW commented on netlogo-gui/src/main/app/AppTabManager.scala:
```diff
@@ -0,0 +1,340 @@
+  // Returns (tabOwner, tabComponent)
+  // tabOwner = the AbstractTabsPanel containing the indexed tab.
+  // tabComponent = the tab in tabOwner referenced by combinedTabIndx.
+  def getTabAtCombinedTabIndx(combinedTabIndx: Int): (AbstractTabsPanel, Component) = {
```
removed
had thoughts of using it to make an iterator over all tabs - if needed for external use

brandesNW commented on netlogo-gui/src/main/app/AppTabManager.scala:
```diff
@@ -0,0 +1,340 @@
+  }
+
+  // Dispatcher to create or end the use of a separate code window
+  // Do nothing if already in desired state
```
removed

brandesNW commented on netlogo-gui/src/main/app/AppTabManager.scala:
```diff
@@ -0,0 +1,340 @@
+      case Some(codeTabsPanel) => {
+        // Move the tabs to the AppTabsPanel (Tabs), retaining order
+        for (n <- 0 until codeTabsPanel.getTabCount) {
+          appTabsPanel.add(codeTabsPanel.getTitleAt(0), codeTabsPanel.getComponentAt(0))
```
I have corrected it to
for (_ <- 0 until codeTabsPanel.getTabCount) {
The code iterates through the codeTabsPanel, removing tabs. Therefore n is not needed as index
At each iteration the first tab is added to the appTabsPanel, and hence is removed from the codeTabsPanel
This way the tabs retain their order in the codeTabsPanel

LaCuneta commented on netlogo-gui/src/main/app/AppTabManager.scala:
```diff
@@ -0,0 +1,326 @@
+      // Move tabs from appTabsPanel to codeTabsPanel.
+      // Iterate starting at last tab so that indexing remains valid when
+      // tabs are removed (add to codeTabsPanel)
+      //val startIndex:Int = appTabsPanel.getTabCount - 1. AAB 10/2020
```
Commented line needs to be removed (but keep the sign/date)

LaCuneta commented on netlogo-gui/src/main/app/CodeTabsPanel.scala:
```diff
@@ -0,0 +1,131 @@
+class CodeTabsPanel(workspace:             GUIWorkspace,
+                    interfaceTab:          InterfaceTab,
+                    externalFileManager:   ExternalFileManager,
+                    val mainCodeTab:           MainCodeTab,
```
Spacing is off here.

LaCuneta commented on netlogo-gui/src/main/app/CodeTabsPanel.scala:
```diff
@@ -0,0 +1,131 @@
+    TabsMenu.tabActions(tabManager).foreach(action => {
+      // Add the accelerator key if any to the input map and action map
+      action.asInstanceOf[MenuAction].accelerator match {
+        case None                =>
```
Still off here.

LaCuneta commented on netlogo-gui/src/main/app/Tabs.scala:
```diff
@@ -60,61 +56,114 @@ class Tabs(val workspace:       GUIWorkspace,
-  val infoTab = new InfoTab(workspace.attachModelDir(_))
-  val codeTab = new MainCodeTab(workspace, this, menu)
-  var externalFileTabs = Set.empty[TemporaryCodeTab]
+  // The value of popOutCodeTab is set in the init method
```
Don't see a resolution here.

LaCuneta commented on netlogo-gui/src/main/app/Tabs.scala:
```diff
@@ -60,61 +56,114 @@ class Tabs(val workspace:       GUIWorkspace,
+      (previousTab.isInstanceOf[TemporaryCodeTab], currentTab.isInstanceOf[TemporaryCodeTab]) match {
+        case (true, false) => saveModelActions foreach menu.offerAction
+        case (false, true) => saveModelActions foreach menu.revokeAction
+        case _             =>
```
Don't see a resolution here.

LaCuneta replied:
Don't see a resolution here.

LaCuneta commented on netlogo-gui/src/main/app/Tabs.scala:
```diff
@@ -60,61 +56,114 @@ class Tabs(val workspace:       GUIWorkspace,
-    OfferSaveExternalsDialog.offer(externalFileTabs filter (_.saveNeeded), this)
+  def handle(e: AboutToCloseFilesEvent) = {
+    OfferSaveExternalsDialog.offer( collection.immutable.Set( externalFileTabs.toSeq: _*).asInstanceOf[Set[TemporaryCodeTab]]
+    filter (_.saveNeeded) , this)
```
Don't see a resolution here.

LaCuneta replied:
I still don't see a resolution here.

LaCuneta commented on netlogo-gui/src/main/app/Tabs.scala:
```diff
@@ -148,28 +196,41 @@ class Tabs(val workspace:       GUIWorkspace,
+    def recolorTab(component: Component, hasError: Boolean): Unit = {
+      tabManager.getTabOwner(component).setForegroundAt(
+        tabManager.getTabOwner(component).indexOfComponent(component),
+        if(hasError) errorColor else null)
```
Don't see a resolution here.

LaCuneta commented on autogen/docs/codetab.md.mustache:
```diff
@@ -5,6 +5,17 @@ to use immediately go in the Command Center; commands you want to save and use
+The Code tab can be opened in a separate window from the main NetLogo window. It can then be moved and resized.
+
+### Ways to separate and rejoin the Code tab
+- The "Code Tab in separate window" checkbox can be used to switch between the joined (part of the main NetLogo window) and separated states.
```
It seems odd to have the parenthetical here, maybe just "...can be used to switch between joined to the main NetLogo window and separated into its own window..."

LaCuneta commented on autogen/docs/codetab.md.mustache:
```diff
@@ -5,6 +5,17 @@ to use immediately go in the Command Center; commands you want to save and use
+- To rejoin a separated Code tab you can use Command + ] on Mac, Control + ]
+number on other operating systems.
+- To rejoin a separated Code tab you can click the close button.
+
```
It seems very odd for this section to come before the table of contents.  I'd recommend putting it after the "Checking for Errors" section, and giving it its own two-`#` header so it also shows up in the ToC.  If you do that, I think you can probably drop the "Ways to separate and rejoin..." header.

LaCuneta commented on autogen/docs/codetab.md.mustache:
```diff
@@ -5,6 +5,8 @@ to use immediately go in the Command Center; commands you want to save and use
 ![[codetab.gif]]
 
+The Code tab can be opened in a separate window from the main NetLogo window. It can then be moved and resized.
+
```
I was assuming this would go into the new code tab section, but if you want to keep it towards the top of the file, I'd move it above the code tab image with the rest of the intro text.

LaCuneta commented on netlogo-gui/src/main/app/Tabs.scala:
```diff
@@ -36,20 +42,11 @@ class Tabs(val workspace:       GUIWorkspace,
 
+  def getTabs = { this }
+
+  def asAbstractTabsPanel = { this.asInstanceOf[AbstractTabsPanel] }
```
I don't understand the use for this.  Isn't `Tabs` already an `AbstractTabsPanel`?  Why do we need to cast it?  Why do we need a method to cast it if we do?

LaCuneta commented on netlogo-gui/src/main/app/CodeTabsPanel.scala:
```diff
@@ -0,0 +1,125 @@
+  val codeTabContainer = new CodeTabContainer(frame, this)
+  val codeTabsPanel = this
+
+  def asAbstractTabsPanel = { this.asInstanceOf[AbstractTabsPanel] }
```
I don't understand the use for this.  Isn't `CodeTabsPanel` already an `AbstractTabsPanel` since it `extends` it?  Why do we need to cast it?  Why do we need a method to cast it if we do?

LaCuneta commented on netlogo-gui/src/main/app/AbstractTabsPanel.scala:
```diff
@@ -0,0 +1,100 @@
+
+  val jframe =  workspace.getFrame.asInstanceOf[JFrame]
+  var tabManager: AppTabManager = null
+  def setTabManager( myTabManager: AppTabManager ) : Unit = {
```
Spacing is off here.

LaCuneta commented on netlogo-gui/src/main/app/AbstractTabsPanel.scala:
```diff
@@ -0,0 +1,100 @@
+    }
+  }
+
+  val jframe =  workspace.getFrame.asInstanceOf[JFrame]
```
Spacing is off here.

LaCuneta commented on netlogo-gui/src/main/app/AppTabManager.scala:
```diff
@@ -0,0 +1,693 @@
+// The terms CombinedIndex or combinedTabIndex refers to the index a tab would have
+// if there were no separate code window. AAB 10/2020
+
+class AppTabManager( val appTabsPanel:          Tabs,
```
Spacing is off here.

LaCuneta commented on netlogo-gui/src/main/app/CodeTabsPanel.scala:
```diff
@@ -0,0 +1,125 @@
+                    interfaceTab:          InterfaceTab,
+                    externalFileManager:   ExternalFileManager,
+                    val mainCodeTab:       MainCodeTab,
+                    val externalFileTabs:  mutable.Set[TemporaryCodeTab])
```
Extra space here.

brandesNW commented:
Hi Jeremy,
I will remove getAppTabsOwner
It was originally there to be parallel to getCodeTabsOwner
but is isn’t necessary.
It’s only used in a method I want to migrate from AbstractCodeTabsPanel in the future.
Aaron
--
Aaron Brandes, Software Developer
Center for Connected Learning and Computer-Based Modeling


From: Jeremy B <notifications@github.com>
Reply-To: NetLogo/NetLogo <reply@reply.github.com>
Date: Wednesday, November 18, 2020 at 10:33 AM
To: NetLogo/NetLogo <NetLogo@noreply.github.com>
Cc: Aaron Andre Brandes <aaron.brandes@northwestern.edu>, Author <author@noreply.github.com>
Subject: Re: [NetLogo/NetLogo] Codetab (#1864)


@LaCuneta commented on this pull request.

________________________________

In netlogo-gui/src/main/app/AppTabManager.scala<https://urldefense.com/v3/__https:/github.com/NetLogo/NetLogo/pull/1864*discussion_r526180188__;Iw!!Dq0X2DkFhyF93HkjWTBQKhk!El6ri1wuIfblFv1mfdDd6Zsm4I4FW8tX5zfE4bM-rVGo307ijsvTad63Gm6XWuwMP-vF5DOmDA$>:

> +  // represented by a scala Option, which has value 'None' when code tabs

+  // are in the Application window and JTabbedPane

+  def setCodeTabsPanelOption(_codeTabsPanelOption: Option[CodeTabsPanel]): Unit = {

+    codeTabsPanelOption = _codeTabsPanelOption

+  }

+

+  // might want access to these owner methods to be only in the app package

+  // Need to carefully decide which methods are private

+  def getCodeTabsOwner = {

+    codeTabsPanelOption match {

+      case None           => appTabsPanel

+      case Some(theValue) => theValue

+    }

+  }

+

+  def getAppTabsOwner = { appTabsPanel }

I'm still not clear on how it's useful to expose appTabsPanel as a method in addition to it being a val on the class (which will never change), but now getAppTabsPanel has also been added which again seems to do the exact same thing? At a minimum there should only be one of these methods, and I think just letting people access appTabsPanel directly is fine.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://urldefense.com/v3/__https:/github.com/NetLogo/NetLogo/pull/1864*discussion_r526180188__;Iw!!Dq0X2DkFhyF93HkjWTBQKhk!El6ri1wuIfblFv1mfdDd6Zsm4I4FW8tX5zfE4bM-rVGo307ijsvTad63Gm6XWuwMP-vF5DOmDA$>, or unsubscribe<https://urldefense.com/v3/__https:/github.com/notifications/unsubscribe-auth/ANSUO2XQNZ3NQVNZT7MC7EDSQPSIDANCNFSM4S2DBXKQ__;!!Dq0X2DkFhyF93HkjWTBQKhk!El6ri1wuIfblFv1mfdDd6Zsm4I4FW8tX5zfE4bM-rVGo307ijsvTad63Gm6XWuwMP-tAeYHqSA$>.

LaCuneta commented on netlogo-gui/src/main/app/Tabs.scala:
```diff
@@ -124,52 +174,79 @@ class Tabs(val workspace:       GUIWorkspace,
-        // item *not* the button) which causes an error in the Code tab the focus gets lost,
-        // so request the focus by a known component 7/18/07
-        requestFocus()
+      case `stableCodeTab` =>
```
Indent is off here.

LaCuneta commented on netlogo-gui/src/main/app/Tabs.scala:
```diff
@@ -189,73 +266,89 @@ class Tabs(val workspace:       GUIWorkspace,
 
-  def addNewTab(name: Filename) = {
-    val tab = new TemporaryCodeTab(workspace, this, name, externalFileManager, fileManager.convertTabAction _, codeTab.smartTabbingEnabled)
+  def addNewNLSTab(name: Filename) = {
```
Don't we call them "External Files" in the rest of the code?  Would it be `addNewExternalFileTab`?

LaCuneta commented on netlogo-gui/src/main/app/TabsMenu.scala:
```diff
@@ -0,0 +1,41 @@
+
+package org.nlogo.app
+
+// TODO i18n lot of work needed here...
```
I might be missing something, I don't see any I18N work to do in this file.  If this comment is kept, sign and date.  If it's from someone else in the past, I think it can just be removed.

LaCuneta commented on netlogo-gui/src/main/app/interfacetab/InterfaceTab.scala:
```diff
@@ -37,7 +37,7 @@ class InterfaceTab(workspace: GUIWorkspace,
   with NlogoPrintable
   with MenuTab {
-
+  def getMonitorManager = monitorManager
```
I don't see this used anywhere?

LaCuneta reviewed (changes requested):
Hey Aaron, a few questions, spacing issues, and such.

brandesNW commented:
See comments below, there are a couple of places I want to look over the code.
I will look again at the tabs location you linked to

--
Aaron Brandes, Software Developer
Center for Connected Learning and Computer-Based Modeling


From: Jeremy B <notifications@github.com>
Reply-To: NetLogo/NetLogo <reply@reply.github.com>
Date: Wednesday, November 18, 2020 at 10:49 AM
To: NetLogo/NetLogo <NetLogo@noreply.github.com>
Cc: Aaron Andre Brandes <aaron.brandes@northwestern.edu>, Author <author@noreply.github.com>
Subject: Re: [NetLogo/NetLogo] Codetab (#1864)


@LaCuneta requested changes on this pull request.

Hey Aaron, a few questions, spacing issues, and such.

________________________________

In netlogo-gui/src/main/app/Tabs.scala<https://urldefense.com/v3/__https:/github.com/NetLogo/NetLogo/pull/1864*discussion_r526187000__;Iw!!Dq0X2DkFhyF93HkjWTBQKhk!ARVOe1rwmb8jcd4Njw2DTn7DSF8-AWT9ezeB6pgaDTmHprugTYJ9aqKN2M3iMVggH6B9s32InA$>:

>

+  def getTabs = { this }

+

+  def asAbstractTabsPanel = { this.asInstanceOf[AbstractTabsPanel] }

I don't understand the use for this. Isn't Tabs already an AbstractTabsPanel? Why do we need to cast it? Why do we need a method to cast it if we do?



will drop. I used this idiom for the tab methods before I moved them because Tabs occurs “naturally” as a JTabbedPane

________________________________

In netlogo-gui/src/main/app/CodeTabsPanel.scala<https://urldefense.com/v3/__https:/github.com/NetLogo/NetLogo/pull/1864*discussion_r526187622__;Iw!!Dq0X2DkFhyF93HkjWTBQKhk!ARVOe1rwmb8jcd4Njw2DTn7DSF8-AWT9ezeB6pgaDTmHprugTYJ9aqKN2M3iMVggH6BLJXVHxw$>:

> +  with LinkParent

+  with org.nlogo.window.LinkRoot {

+

+  locally {

+    addChangeListener(this)

+  }

+

+  // frame is the App's AppFrame, (treated as a java.awt.Frame) AAB 10/2020

+  val frame = workspace.getFrame

+

+  // CodeTabContainer contains the CodeTabsPanel and is owned by frame

+  // It is currently implemented as a JDialog. AAB 10/2020

+  val codeTabContainer = new CodeTabContainer(frame, this)

+  val codeTabsPanel = this

+

+  def asAbstractTabsPanel = { this.asInstanceOf[AbstractTabsPanel] }

I don't understand the use for this. Isn't CodeTabsPanel already an AbstractTabsPanel since it extends it? Why do we need to cast it? Why do we need a method to cast it if we do?

will drop. I used this idiom for the tab methods before I moved them because CodeTabsPanel occurs “naturally” as a JTabbedPane



________________________________

In netlogo-gui/src/main/app/AbstractTabsPanel.scala<https://urldefense.com/v3/__https:/github.com/NetLogo/NetLogo/pull/1864*discussion_r526188141__;Iw!!Dq0X2DkFhyF93HkjWTBQKhk!ARVOe1rwmb8jcd4Njw2DTn7DSF8-AWT9ezeB6pgaDTmHprugTYJ9aqKN2M3iMVggH6DTgEjqPg$>:

> +  locally {

+    setOpaque(false)

+    setFocusable(false)

+    if (System.getProperty("os.name").startsWith("Mac")) {

+      try {

+        val ui = Class.forName("org.nlogo.app.MacTabbedPaneUI").newInstance.asInstanceOf[ComponentUI]

+        setUI(ui)

+      } catch {

+        case e: ClassNotFoundException =>

+      }

+    }

+  }

+

+  val jframe =  workspace.getFrame.asInstanceOf[JFrame]

+  var tabManager: AppTabManager = null

+  def setTabManager( myTabManager: AppTabManager ) : Unit = {

Spacing is off here.

________________________________

In netlogo-gui/src/main/app/AbstractTabsPanel.scala<https://urldefense.com/v3/__https:/github.com/NetLogo/NetLogo/pull/1864*discussion_r526188303__;Iw!!Dq0X2DkFhyF93HkjWTBQKhk!ARVOe1rwmb8jcd4Njw2DTn7DSF8-AWT9ezeB6pgaDTmHprugTYJ9aqKN2M3iMVggH6BG6BzjWA$>:

> +  extends JTabbedPane(SwingConstants.TOP) {

+

+  locally {

+    setOpaque(false)

+    setFocusable(false)

+    if (System.getProperty("os.name").startsWith("Mac")) {

+      try {

+        val ui = Class.forName("org.nlogo.app.MacTabbedPaneUI").newInstance.asInstanceOf[ComponentUI]

+        setUI(ui)

+      } catch {

+        case e: ClassNotFoundException =>

+      }

+    }

+  }

+

+  val jframe =  workspace.getFrame.asInstanceOf[JFrame]

Spacing is off here.

________________________________

In netlogo-gui/src/main/app/AppTabManager.scala<https://urldefense.com/v3/__https:/github.com/NetLogo/NetLogo/pull/1864*discussion_r526188529__;Iw!!Dq0X2DkFhyF93HkjWTBQKhk!ARVOe1rwmb8jcd4Njw2DTn7DSF8-AWT9ezeB6pgaDTmHprugTYJ9aqKN2M3iMVggH6AqMVGf4A$>:

> +// code controlling tabs.

+// When there is no separate code window, all tabs belong to Tabs

+// (the value of appTabsPanel in AppTabManager)

+// The tab order is non-CodeTabs followed by CodeTabs.

+// The non-CodeTab order is InterfaceTab, InfoTab, any user-created non-CodeTabs.

+// The MainCodeTab is always the first CodeTab, the other CodeTabs are the

+// TemporaryCodeTab which is used for included files, and the

+// ExtendedCodeTab which is for use outside the NetLogo code base.

+//

+// When a separate code window exists, all the CodeTabs belong to the CodeTabsPanel.

+// They retain the same relative order they would if they all belonged to Tabs.

+//

+// The terms CombinedIndex or combinedTabIndex refers to the index a tab would have

+// if there were no separate code window. AAB 10/2020

+

+class AppTabManager( val appTabsPanel:          Tabs,

Spacing is off here.

________________________________

In netlogo-gui/src/main/app/CodeTabsPanel.scala<https://urldefense.com/v3/__https:/github.com/NetLogo/NetLogo/pull/1864*discussion_r526189447__;Iw!!Dq0X2DkFhyF93HkjWTBQKhk!ARVOe1rwmb8jcd4Njw2DTn7DSF8-AWT9ezeB6pgaDTmHprugTYJ9aqKN2M3iMVggH6BUmBR7FA$>:

> +import org.nlogo.window.GUIWorkspace

+import org.nlogo.window.Event

+import org.nlogo.window.Event.LinkParent

+import org.nlogo.window.Events._

+

+// When a separate code tab window is created, an instance of this class owns the CodeTabs.

+// When there is no separate code tab window, no such instance exists, and all

+// CodeTabs belong to Tabs. [ Thinking of Tabs as AppTabsPanel makes the parallels between

+// CodeTabsPanel and Tabs clearer - both are JTabbedPanes that contain and manage tabs.]

+// CodeTabsPanel and Tabs are both instances of AbstractTabsPanel, which implements their shared behavior. AAB 10/2020

+

+class CodeTabsPanel(workspace:             GUIWorkspace,

+                    interfaceTab:          InterfaceTab,

+                    externalFileManager:   ExternalFileManager,

+                    val mainCodeTab:       MainCodeTab,

+                    val externalFileTabs:  mutable.Set[TemporaryCodeTab])

Extra space here.

Do you mean after the colons?

________________________________

In netlogo-gui/src/main/app/Tabs.scala<https://urldefense.com/v3/__https:/github.com/NetLogo/NetLogo/pull/1864*discussion_r526190961__;Iw!!Dq0X2DkFhyF93HkjWTBQKhk!ARVOe1rwmb8jcd4Njw2DTn7DSF8-AWT9ezeB6pgaDTmHprugTYJ9aqKN2M3iMVggH6DJs3iX5g$>:

>        if (e.error != null) setSelectedIndex(0)

       recolorTab(interfaceTab, e.error != null)

     }



     // recolor tabs

     e.sourceOwner match {

-      case `codeTab` =>

-        // on null error, clear all errors, as we only get one event for all the files

-        if (e.error == null)

-          clearErrors()

-        else {

-          setSelectedComponent(codeTab)

-          recolorTab(codeTab, true)

-        }

-        // I don't really know why this is necessary when you delete a slider (by using the menu

-        // item *not* the button) which causes an error in the Code tab the focus gets lost,

-        // so request the focus by a known component 7/18/07

-        requestFocus()

+      case `stableCodeTab` =>

Indent is off here.

________________________________

In netlogo-gui/src/main/app/Tabs.scala<https://urldefense.com/v3/__https:/github.com/NetLogo/NetLogo/pull/1864*discussion_r526191896__;Iw!!Dq0X2DkFhyF93HkjWTBQKhk!ARVOe1rwmb8jcd4Njw2DTn7DSF8-AWT9ezeB6pgaDTmHprugTYJ9aqKN2M3iMVggH6BPq3PDxA$>:

>

-  def addNewTab(name: Filename) = {

-    val tab = new TemporaryCodeTab(workspace, this, name, externalFileManager, fileManager.convertTabAction _, codeTab.smartTabbingEnabled)

+  def addNewNLSTab(name: Filename) = {

Don't we call them "External Files" in the rest of the code? Would it be addNewExternalFileTab?

sure

________________________________

In netlogo-gui/src/main/app/TabsMenu.scala<https://urldefense.com/v3/__https:/github.com/NetLogo/NetLogo/pull/1864*discussion_r526192751__;Iw!!Dq0X2DkFhyF93HkjWTBQKhk!ARVOe1rwmb8jcd4Njw2DTn7DSF8-AWT9ezeB6pgaDTmHprugTYJ9aqKN2M3iMVggH6A54lc8cw$>:

> @@ -0,0 +1,41 @@

+// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

+

+package org.nlogo.app

+

+// TODO i18n lot of work needed here...

I might be missing something, I don't see any I18N work to do in this file. If this comment is kept, sign and date. If it's from someone else in the past, I think it can just be removed.

Previous code reviewer missed that

________________________________

In netlogo-gui/src/main/app/interfacetab/InterfaceTab.scala<https://urldefense.com/v3/__https:/github.com/NetLogo/NetLogo/pull/1864*discussion_r526193218__;Iw!!Dq0X2DkFhyF93HkjWTBQKhk!ARVOe1rwmb8jcd4Njw2DTn7DSF8-AWT9ezeB6pgaDTmHprugTYJ9aqKN2M3iMVggH6AI29YV7g$>:

> @@ -37,7 +37,7 @@ class InterfaceTab(workspace: GUIWorkspace,

   with SwitchedTabsEvent.Handler

   with NlogoPrintable

   with MenuTab {

-

+  def getMonitorManager = monitorManager

I don't see this used anywhere?

gone

________________________________

In netlogo-gui/src/main/app/Tabs.scala<https://urldefense.com/v3/__https:/github.com/NetLogo/NetLogo/pull/1864*discussion_r526190212__;Iw!!Dq0X2DkFhyF93HkjWTBQKhk!ARVOe1rwmb8jcd4Njw2DTn7DSF8-AWT9ezeB6pgaDTmHprugTYJ9aqKN2M3iMVggH6CH8i8U-w$>:

> +      currentTab = getSelectedComponent

+      tabManager.setCurrentTab(currentTab)

+

+      previousTab match {

+        case mt: MenuTab => mt.activeMenuActions foreach menu.revokeAction

+        case _ =>

+      }

+      currentTab match {

+        case mt: MenuTab => mt.activeMenuActions foreach menu.offerAction

+        case _ =>

+      }

+      // The correctness of this logic needs investigation - AAB 10/2020

+      (previousTab.isInstanceOf[TemporaryCodeTab], currentTab.isInstanceOf[TemporaryCodeTab]) match {

+        case (true, false) => saveModelActions foreach menu.offerAction

+        case (false, true) => saveModelActions foreach menu.revokeAction

+        case _             =>

Don't see a resolution here.

Probably don’t need comment, but this is code I would like to look over with you

________________________________

In netlogo-gui/src/main/app/Tabs.scala<https://urldefense.com/v3/__https:/github.com/NetLogo/NetLogo/pull/1864*discussion_r526188711__;Iw!!Dq0X2DkFhyF93HkjWTBQKhk!ARVOe1rwmb8jcd4Njw2DTn7DSF8-AWT9ezeB6pgaDTmHprugTYJ9aqKN2M3iMVggH6BEjH_fYA$>:

>

-  def handle(e: AboutToCloseFilesEvent) =

-    OfferSaveExternalsDialog.offer(externalFileTabs filter (_.saveNeeded), this)

+  def handle(e: AboutToCloseFilesEvent) = {

+    OfferSaveExternalsDialog.offer( collection.immutable.Set( externalFileTabs.toSeq: _*).asInstanceOf[Set[TemporaryCodeTab]]

+    filter (_.saveNeeded) , this)

I still don't see a resolution here.

Lets’ look at this

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://urldefense.com/v3/__https:/github.com/NetLogo/NetLogo/pull/1864*pullrequestreview-533575613__;Iw!!Dq0X2DkFhyF93HkjWTBQKhk!ARVOe1rwmb8jcd4Njw2DTn7DSF8-AWT9ezeB6pgaDTmHprugTYJ9aqKN2M3iMVggH6CtX4SO4Q$>, or unsubscribe<https://urldefense.com/v3/__https:/github.com/notifications/unsubscribe-auth/ANSUO2QKVX2OYVIZHCNEYELSQPUHRANCNFSM4S2DBXKQ__;!!Dq0X2DkFhyF93HkjWTBQKhk!ARVOe1rwmb8jcd4Njw2DTn7DSF8-AWT9ezeB6pgaDTmHprugTYJ9aqKN2M3iMVggH6DRgwjF0Q$>.
